internal/hcp/internal/controllers/link/controller.go

```diff
@@ -80,3 +93,65 @@ func (r *linkReconciler) Reconcile(ctx context.Context, rt controller.Runtime, r
 
 	return nil
 }
+
+type linkInitializer struct {
+	cloudConfig config.CloudConfig
+}
+
+func (i *linkInitializer) Initialize(ctx context.Context, rt controller.Runtime) error {
+	if !i.cloudConfig.IsConfigured() {
+		return nil
+	}
+
+	// Construct a link resource to reflect the configuration
+	data, err := anypb.New(&pbhcp.Link{
+		ResourceId:   i.cloudConfig.ResourceID,
+		ClientId:     i.cloudConfig.ClientID,
+		ClientSecret: i.cloudConfig.ClientSecret,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Create the link resource for a configuration-based link
+	w := retry.Waiter{
+		MinWait: 1 * time.Second,
+		MaxWait: 15 * time.Second,
+		Jitter:  retry.NewJitter(50),
+	}
+	retryCtx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+	var writeErr error
+	for {
+		_, writeErr = rt.Client.Write(ctx,
+			&pbresource.WriteRequest{
+				Resource: &pbresource.Resource{
+					Id: &pbresource.ID{
+						Name: types.LinkName,
+						Type: pbhcp.LinkType,
+					},
+					Metadata: map[string]string{
+						MetadataSourceKey: MetadataSourceConfig,
+					},
+					Data: data,
+				},
+			},
+		)
+		if writeErr != nil {
+			// Retry on error
+			rt.Logger.Trace("error initializing link, retrying", "error", writeErr)
+			if err := w.Wait(retryCtx); err != nil {
+				// Retry timed out, break out of retry loop
```

Comment:
Assuming this break then calls [this](https://github.com/hashicorp/consul/pull/20138/files#diff-45b952ceebe565e9d18a06a30c991e0e0f6303b86866958195d6dc462be72669R152-R154)? If so, can we just do the return here?